allow remote torchscript call to itself

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32959 jit pickling rref
* **#32990 allow remote torchscript call to itself**

right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

zhaojuanmao committed:
allow remote torchscript call to itself

right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "allow remote torchscript call to itself"


right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "allow remote torchscript call to itself"


right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

[ghstack-poisoned]

zhaojuanmao commented:
lint and test failures are not relevant

xush6528 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -119,11 +120,10 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
-      return wrap(
-          RemoteRet(scriptRemoteCall.retRRefId(), scriptRemoteCall.retForkId())
-              .toMessage());
+      if (rrefId != forkId) {
```
I feel we need some comment here to explain the difference of 2 cases

1) In what case, rrefId == forkId
2) why, in this case, don't need to add a fork

xush6528 commented on torch/csrc/distributed/rpc/torchscript_functions.cpp:
```diff
@@ -72,25 +70,61 @@ std::shared_ptr<UserRRef> remoteTorchscript(
+          rr->rrefId() == rr->forkId(),
+          "Expecting an OwnerRRef as RemoteRet but got a fork.");
+      auto& refCtx = RRefContext::getInstance();
+      refCtx.delForkOfOwner(rr->rrefId(), rr->rrefId());
```
I prefer to refactor `finishCreatingOwnerRRef()` into 
`finishCreatingOwnerRRef()` (in rref_ccontetx.cpp callback namespace) and `finishCreatingOwnerRRefWithGIL()` (in python_functions) to reduce duplicate code.

zhaojuanmao replied:
future call back schema is defined as void function(const message&, FutureError&), finishCreatingOwnerRRefWithGIL() can not call finishCreatingOwnerRRef(), as it needs finishCreatingOwnerRRef() return deletedRRef?

xush6528 replied:
Maybe this way?

std::shared_ptr<RRef> finishCreatingOwnerRRef() {

}

For Python remote call
```
    fm->addCallback([](const Message& message,
                       const c10::optional<utils::FutureError>& futErr) {
      auto deletedRRef = callback::finishCreatingOwnerRRef(message, futErr);
     if (deletedRRef && deletedRRef->isPyObj()) {
          AutoGil gil;
          deletedRRef.reset()
     }
```

For Script remote call
```
    fm->addCallback([](const Message& message,
                       const c10::optional<utils::FutureError>& futErr) {
      callback::finishCreatingOwnerRRef(message, futErr);
```

zhaojuanmao replied:
sounds good!

zhaojuanmao committed:
Update on "allow remote torchscript call to itself"


right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/torchscript_functions.cpp:
```diff
@@ -72,25 +70,55 @@ std::shared_ptr<UserRRef> remoteTorchscript(
+    // python_functions.py, the difference is that ownerRRefPtr created here
+    // always holds an IValue, no need to explicitly hold GIL and clear
+    // deletedRRef. Not calling finishCreatingOwnerRRef() because it is not
+    // necessary to introduce python dependency in this file.
```
Need to update comment?

xush6528 commented:
Looks good to me. @mrshenli Any more comment?

zhaojuanmao committed:
Update on "allow remote torchscript call to itself"


right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/torchscript_functions.cpp:
```diff
@@ -72,25 +70,50 @@ std::shared_ptr<UserRRef> remoteTorchscript(
+    return userRRefPtr;
+  } else {
+    auto ownerRRefPtr = ctx.createOwnerRRef(returnType);
+    // prevent this owner RRef be deleted due to other forks
```
be -> from being

mrshenli commented on torch/csrc/distributed/rpc/torchscript_functions.cpp:
```diff
@@ -72,25 +70,50 @@ std::shared_ptr<UserRRef> remoteTorchscript(
+        nullptr);
 
-  return userRRefPtr;
+    fm->addCallback([](const Message& message,
```
can this be 

```c++
fm->addCallback(callback::finishCreatingOwnerRRef)?
```

mrshenli reviewed (commented):
The changes look good to me, but there are a lot of tests with irrelevant failures. Shall we rebase and clear the test to get meaningful signals? Thanks!

zhaojuanmao committed:
Update on "allow remote torchscript call to itself"


right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "allow remote torchscript call to itself"


right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

[ghstack-poisoned]

zhaojuanmao commented:
test failures are not related

mrshenli reviewed (approved):
Thanks for working on this! Stamp to unblock.

Lint and test failures are irrelevant.

zhaojuanmao committed:
Update on "allow remote torchscript call to itself"


right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

[ghstack-poisoned]
